app/javascript/utilities/viewport.js

```diff
@@ -7,18 +7,35 @@
  *
  * @param {object} element - The HTML element to check
  * @param {number} [offsetTop=0] - Part of the screen to ignore counting from the top
- *
+ * @param {boolean} [isPartiallyVisible=false] - A boolean to flip the check between partial or completely visible in the viewport
  * @returns {boolean} isInViewport - true if the element is visible in the viewport
  */
-export function isInViewport(element, offsetTop = 0) {
+export function isInViewport(
+  element,
+  offsetTop = 0,
+  isPartiallyVisible = false,
+) {
   const boundingRect = element.getBoundingClientRect();
   const clientHeight =
     window.innerHeight || document.documentElement.clientHeight;
   const clientWidth = window.innerWidth || document.documentElement.clientWidth;
-  return (
-    boundingRect.top >= offsetTop &&
-    boundingRect.left >= 0 &&
-    boundingRect.bottom <= clientHeight &&
-    boundingRect.right <= clientWidth
-  );
+  const topIsInViewport =
+    boundingRect.top <= clientHeight && boundingRect.top >= offsetTop;
+  const rightIsInViewport =
+    boundingRect.right >= 0 && boundingRect.right <= clientWidth;
+  const bottomIsInViewport =
+    boundingRect.bottom >= offsetTop && boundingRect.bottom <= clientHeight;
+  const leftIsInViewport =
+    boundingRect.left <= clientWidth && boundingRect.left >= 0;
+  const topIsOutOfViewport = boundingRect.top <= offsetTop;
+  const bottomIsOutOfViewport = boundingRect.bottom >= clientHeight;
+  const elementSpansEntireViewport =
+    topIsOutOfViewport && bottomIsOutOfViewport;
+  return isPartiallyVisible
+    ? (topIsInViewport || bottomIsInViewport || elementSpansEntireViewport) &&
+        (leftIsInViewport || rightIsInViewport)
+    : topIsInViewport &&
+        bottomIsInViewport &&
+        leftIsInViewport &&
+        rightIsInViewport;
```

Comment:
(nitpick / not-blocking) - I think in this case the ternary makes it a little harder to read as we have so many conditionals chained together. I'd suggest splitting out into an `if` to make it more readable:

```js
 if (allowPartialVisibility) {
    return (
      (topIsInViewport || bottomIsInViewport || elementSpansEntireViewport) &&
      (leftIsInViewport || rightIsInViewport)
    );
  }

  return (
    topIsInViewport &&
    bottomIsInViewport &&
    leftIsInViewport &&
    rightIsInViewport
  );
```

Reply:
Thanks! Yes, the ternary demanded more cognitive load. This looks a lot better.